src/OrchardCore.Modules/OrchardCore.Users/Startup.cs

```diff
@@ -491,6 +500,18 @@ public override void ConfigureServices(IServiceCollection services)
         {
             services.AddScoped<IDisplayDriver<User>, CustomUserSettingsDisplayDriver>();
             services.AddScoped<IPermissionProvider, CustomUserSettingsPermissions>();
+
+            services.AddTransient<IDeploymentSource, CustomUserSettingsDeploymentSource>();
+            services.AddSingleton<IDeploymentStepFactory>(new DeploymentStepFactory<CustomUserSettingsDeploymentStep>());
+            services.AddScoped<IDisplayDriver<DeploymentStep>, CustomUserSettingsDeploymentStepDriver>();
+        }
+    }
+
+    public class UserDeploymentStartup : StartupBase {
+        public override void ConfigureServices(IServiceCollection services) {
+            services.AddTransient<IDeploymentSource, AllUsersDeploymentSource>();
+            services.AddSingleton<IDeploymentStepFactory>(new DeploymentStepFactory<AllUsersDeploymentStep>());
```

Comment:
why `new`? either register the type or use the action to return a new instance.

Reply:
Try changing it to `services.AddSingleton<IDeploymentStepFactory>(sp => new DeploymentStepFactory<AllUsersDeploymentStep>());`

Reply:
I think that's okay for now, it is used like this in other places because it is a stateless service that doesn't need services injection. A factory is only used to build unititialized step of a given type inheriting from `DeploymentStep` and with a parameterless ctor (`new()` constraint).

We could have registered  a generic type in one place, from memory like this

    services.AddSingleton(typeof(IDeploymentStepFactory<>), typeof(DeploymentStepFactory<>));

In the `OC.Secrets` I'm working on, in place of registering in the DI, the related manager builds internally generic `Activators` as we use for example to activate content Parts and Fields.